Adding CloudEvents samples for ASB and SQS (#7933)
Particular/docs.particular.net: transports/cloudevents.md

---
title: CloudEvents support
summary: Explains how to consume CloudEvents in transports
component: Core
reviewed: 2025-12-16
versions: '[10,]'
---

> [!WARNING]

> > This is an experimental feature and might change in the future.

Reviewer: ```suggestion
> This is an experimental feature and, as such, is subject to changes.
```

This guideline shows how to receive [CloudEvents](https://cloudevents.io/) in the NServiceBus endpoints.

## Supported formats

This section describes [CloudEvents formats](https://github.com/cloudevents/spec/tree/v1.0.2/cloudevents/formats) supported by NServiceBus.

### JSON

NServiceBus supports [JSON format](https://github.com/cloudevents/spec/blob/v1.0.2/cloudevents/formats/json-format.md#3-envelope) only. It does not support [JSON Batch format](https://github.com/cloudevents/spec/blob/v1.0.2/cloudevents/formats/json-format.md#4-json-batch-format).

The implementation relies on [System.Text.Json](https://learn.microsoft.com/en-us/dotnet/api/system.text.json?view=net-10.0) package to parse the JSON content.

## Supported bindings

This section describes [CloudEvents bindings](https://github.com/cloudevents/spec/tree/v1.0.2/cloudevents/bindings) supported by NServiceBus.

### Strutured Content Mode for HTTP and AMQP

> NServiceBus supports [HTTP Structured Content Mode](https://github.com/cloudevents/spec/blob/v1.0.2/cloudevents/bindings/http-protocol-binding.md#32-structured-content-mode) and [AMQP Structured Content Mode](https://github.com/cloudevents/spec/blob/v1.0.2/cloudevents/bindings/amqp-protocol-binding.md#32-structured-content-mode). Two implementation modes are provided: Strict and Permissive.

Reviewer: ```suggestion
NServiceBus supports [HTTP Structured](https://github.com/cloudevents/spec/blob/v1.0.2/cloudevents/bindings/http-protocol-binding.md#32-structured-content-mode) and [AMQP Structured](https://github.com/cloudevents/spec/blob/v1.0.2/cloudevents/bindings/amqp-protocol-binding.md#32-structured-content-mode) content modes. For each content mode, two implementations are provided:

- Strict
- Permissive
```

#### Strict mode

> To recognize the CloudEvents message, the `Content-Type` header must be equal to or contain `application/cloudevents+json`. The implementation will not attempt to deserialize the payload if the `Content-Type` header does not meet the requirements.

Reviewer: ```suggestion
In Strict mode, the CloudEvents message `Content-Type` header must be equal to or contain `application/cloudevents+json`. The implementation will not attempt to deserialize the payload if the `Content-Type` header does not meet the requirements.
```

Fields [`id`](https://github.com/cloudevents/spec/blob/v1.0.2/cloudevents/spec.md#id), [`source`](https://github.com/cloudevents/spec/blob/v1.0.2/cloudevents/spec.md#source-1), [`type`](https://github.com/cloudevents/spec/blob/v1.0.2/cloudevents/spec.md#type), and at least one of `data` and `data_base64` must be present. The [`specversion`](https://github.com/cloudevents/spec/blob/v1.0.2/cloudevents/spec.md#specversion) is not required to be present or to equal to `1.0`.

> 

Reviewer: ```suggestion
```

The [HTTP Content-Type](https://github.com/cloudevents/spec/blob/v1.0.2/cloudevents/bindings/http-protocol-binding.md#311-http-content-type) and [AMQP Content-Type](https://github.com/cloudevents/spec/blob/v1.0.2/cloudevents/bindings/amqp-protocol-binding.md#321-amqp-content-type) headers are not validated.

## Configuration

This section describes the configuration options.

### Enabling the configuration

To enable the CloudEvents support:

snippet: sqs-cloudevents-configuration

### Type mapping

> `TypeMappings` configure how to match the incoming messages with the class definitions:

Reviewer: ```suggestion
`TypeMappings` configure how to match the incoming message content-type value with the class definition used in the NServiceBus message handler:
```

snippet: sqs-cloudevents-typemapping

### Enabled unwrappers

> The `EnvelopeUnwrappers` property contains the list of enabled modes. By default, both Binary Content Mode and Structured Content Mode in Strict Mode are enabled.

Reviewer: Something is off with the word "mode," I cannot pinpoint it, but the first sentence doesn't read well. Should we move away from mode to something like envelope format? In which case, that first sentence would be:

> The `EnvelopeUnwrappers` property lists the enabled envelope formats.
